xdsl/dialects/builtin.py

```diff
@@ -1004,13 +1004,72 @@ def unpack(
         return tuple(FloatAttr(value, type) for value in type.unpack(buffer, num))
 
 
+ComplexElementT = TypeVar(
+    "ComplexElementT", bound=IntegerType | AnyFloat, default=IntegerType | AnyFloat
+)
+
+
 @irdl_attr_definition
-class ComplexType(ParametrizedAttribute, BuiltinAttribute, TypeAttribute):
+class ComplexType(
+    Generic[ComplexElementT],
+    StructPackableType[float | int],
+    ParametrizedAttribute,
+    BuiltinAttribute,
+    TypeAttribute,
+):
     name = "complex"
-    element_type: ParameterDef[IntegerType | AnyFloat]
+    element_type: ParameterDef[ComplexElementT]
+
+    def __init__(self, element_type: AttributeCovT):
+        super().__init__([element_type])
+
+    def get_element_type(self) -> ComplexElementT:
+        return self.element_type
+
+    @property
+    def size(self) -> int:
+        return 2 * self.element_type.size
+
+    @property
+    def format(self) -> str:
+        return f"<2{self.element_type.format[1]}"
+
+    def iter_unpack(self, buffer: ReadableBuffer, /) -> Iterator[float | int]:
+        return (values[0] for values in struct.iter_unpack(self.format, buffer))
+
+    def unpack(self, buffer: ReadableBuffer, num: int, /) -> tuple[float | int, ...]:
+        fmt = self.format[0] + str(2 * num) + self.format[2:]
+        return struct.unpack(fmt, buffer)
+
+    @overload
+    def pack_into(
+        self, buffer: WriteableBuffer, offset: int, value: float | int
+    ) -> None: ...
+
+    @overload
+    def pack_into(
+        self,
+        buffer: WriteableBuffer,
+        offset: int,
+        value: tuple[float, float] | tuple[int, int],
+    ) -> None: ...
+
+    def pack_into(
```

Comment:
Should be able to simplify this with the struct packable type being `complex | tuple[int, int]`